### PR TITLE
Add cc bcc csv comma delimited V2

### PR DESCRIPTION
**Pull request by holstbone**
@liamlynch-data - I created a clean, new pull request here.
(original PR [here](https://github.com/dataiku/dss-plugin-sendmail/pull/20) )

NOTES:

I did NOT incorporate changing to `SMTP.send_message(msg)`, for two reasons:

1. The `sendmail` method was selected by the Plugin dev team, so updates should probably best be led by them.
2. There may be some ramifications to how attachments are managed.

Also, I did not incorporate hard coded CC or BCC fields, primarily because the UI/UX gets really messy. Since "Recipient" field is always pulled from the dataset, then the best UI/UX is to also always pull the "cc" and "bcc" from the dataset as well.

### CODE REVIEW

File: custom-recipes/send-mails-from-contacts-dataset/recipe.py [hunk 1]
```diff
@@ -120,9 +123,21 @@ def parse_recipients(recipients):
...
+        raise AttributeError("The column you specified (%s) was not found." % arg)
+    
+# Validation part 3 - validate that recipient, cc, and bcc are string datatypes
+people_schema = people.read_schema()
+
+for check in [recipient_column, cc_column, bcc_column]:
+    for item in people_schema:
+        if item['name'] == check:
+            if item['type'] != 'string':
+                raise AttributeError("The column you specified for (%s) was not datatype string." % item['name'])
```
Comment (liamlynch-data) [inline custom-recipes/send-mails-from-contacts-dataset/recipe.py]: 🔧  the validation code is getting a bit messy and inconsistent, and I don't think we really need a third validation section to do this check - perhaps it'd be best to create a function like this:
```python
def confirm_string_column(column_name, schema):
    column = next(filter(lambda col: col['name'] == column_name, schema), None)
    if not column:
          raise AttributeError("The column you specified (%s) was not found." % column_name)
    if column['type'] != 'string':
        raise AttributeError("The column you specified (%s) was not datatype string." % column_name)
 ```
 Then rewrite the validation step 2 along these lines, checking everything the same way:
 
 ```python
 # Validation part 2 - when necessary, check the columns given are present as string columns in the contacts (people) dataset
people_schema = people.read_schema()
for arg in ['subject', 'body']:
    if not globals()["use_" + arg + "_value"]:
        confirm_string_column(globals()[arg + "_column"], people_schema)       

if not channel_has_sender and not use_sender_value:    
    confirm_string_column(sender_column, people_schema)

for arg in [recipient_column, cc_column, bcc_column]:
    if not arg:
        # recipient_column would have previously failed in Validation Part 1
        pass
    else:
        confirm_string_column(arg, people_schema)
        
```

File: python-lib/dku_attachment_handling.py [hunk 2]
```diff
@@ -58,17 +58,22 @@ def build_attachment_files(attachment_datasets, attachment_type, apply_coloring_
...
             format_params = {"applyColoring": True}
+    elif attachment_type == "csv_comma":
+        request_fmt = "csv"
+        format_params={"style": "excel", "separator": ",", "quoteChar" : "\"", "parseHeaderRow": True}
     else:
         request_fmt = "tsv-excel-header"
 
     # Prepare attachments
     attachment_files = []
+    for attachment_ds in attachment_datasets:
```
Comment (liamlynch-data) [inline python-lib/dku_attachment_handling.py]: 💥 This double line breaks things - I guess it was an accident

File: python-lib/dku_attachment_handling.py [hunk 3]
```diff
@@ -58,17 +58,22 @@ def build_attachment_files(attachment_datasets, attachment_type, apply_coloring_
         request_fmt = "excel"
         if apply_coloring_excel and supports_messaging_channels_and_conditional_formatting(dataiku.api_client()):
             format_params = {"applyColoring": True}
+    elif attachment_type == "csv_comma":
+        request_fmt = "csv"
+        format_params={"style": "excel", "separator": ",", "quoteChar" : "\"", "parseHeaderRow": True}
     else:
```
Comment (liamlynch-data) [inline python-lib/dku_attachment_handling.py:66]: ⛏️  nickpick: For the else, I would still indicate (in code or just with a comment) that we expect `attachment_type == "csv"` here. In case we add any more options, it just makes it clearer.

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,5 +1,9 @@
 # Changelog
 
+## [Version 1.0.4]() - Custom feature release 2025-11
```
Comment (liamlynch-data) [inline CHANGELOG.md]: I think we can just call it "Feature release" - and also add a link to the tag like for the others (it doesn't exist yet but it will:
https://github.com/dataiku/dss-plugin-sendmail/releases/tag/v1.0.4

File: python-lib/dku_email_client.py [hunk 5]
```diff
@@ -131,30 +142,37 @@ def attachments_to_mime(self, attachment_files):
...
         :param email_body: body of either plain text or html, str
         :param attachment_mimes, list of MIMEBase
         """
         msg = MIMEMultipart()
         msg["From"] = sender
         msg["To"] = ",".join(recipients)
+
+        msg["Cc"] = ",".join(cc_recipients)
+        all_recipients = recipients + cc_recipients + bcc_recipients
+
```
Comment (liamlynch-data) [inline python-lib/dku_email_client.py:167]: ⛏️ nitpick: maybe add a comment on how the BCC field is implicitly handled

**Review by liamlynch-data — CHANGES_REQUESTED** [review]
That's much better - thanks for the changes!

Re:
I did NOT incorporate changing to SMTP.send_message(msg)
Ok no worries - perhaps we'll change it later, it doesn't matter too much

Also, I did not incorporate hard coded CC or BCC fields
Not really an issue either

There are a couple more things to address here - if you can make those changes then I'll do a full test on it (I gave it a quick test already, seems fine once I fixed that double line issue).

File: custom-recipes/send-mails-from-contacts-dataset/recipe.py (hunk 1 above)
Comment (holstbone) [reply custom-recipes/send-mails-from-contacts-dataset/recipe.py]: Addressed!

File: python-lib/dku_attachment_handling.py (hunk 2 above)
Comment (holstbone) [reply python-lib/dku_attachment_handling.py]: Yeah ... opps.  Fixed!

File: python-lib/dku_attachment_handling.py (hunk 3 above)
Comment (holstbone) [reply python-lib/dku_attachment_handling.py:66]: Added #comment

File: python-lib/dku_email_client.py (hunk 5 above)
Comment (holstbone) [reply python-lib/dku_email_client.py:167]: Added #comment

**Comment by holstbone** [comment]
@liamlynch-data - Just made final adjustments!

File: CHANGELOG.md (hunk 4 above)
Comment (holstbone) [reply CHANGELOG.md]: Adjusted!

File: CHANGELOG.md [hunk 6]
```diff
@@ -1,5 +1,9 @@
 # Changelog
 
+## [Version 1.0.4]() - Feature release - 2025-11
```
Comment (liamlynch-data) [inline CHANGELOG.md]: Needs a minor tweak here
```suggestion
## [Version 1.0.4](https://github.com/dataiku/dss-plugin-sendmail/releases/tag/v1.0.4)  - Feature release - 2025-12
```

Comment (holstbone) [reply CHANGELOG.md]: Corrected

**Comment by holstbone** [comment]
All changes from the sendmail_changes.patch have now been incorporated!

File: custom-recipes/send-mails-from-contacts-dataset/recipe.py [hunk 7]
```diff
@@ -34,25 +34,33 @@ def to_real_channel_id(channel_id):
 def does_channel_have_sender(channel_id):
     return channel_id is not None and channel_id.endswith(SENDER_SUFFIX)
 
-# Takes a string and returns a list of one or more address values
+# Takes a string and returns a list of address values, could be empty - any no blank/empty strings are filtered out
```
Comment (liamlynch-data) [inline custom-recipes/send-mails-from-contacts-dataset/recipe.py:37]: Explaining patch changes: I added code to filter out blank strings in parce_recipients() as they can can cause problems later and we can just skip them.

File: custom-recipes/send-mails-from-contacts-dataset/recipe.py [hunk 8]
```diff
@@ -118,25 +126,37 @@ def confirm_string_column(column_name, schema):
...
+if channel_has_sender:
+    is_sender_present = True
+else:
+    if use_sender_value:
+        is_sender_present = bool(sender_value)
+    else:
+        is_sender_present = bool(sender_column)
+if not is_sender_present:
+    raise AttributeError("No value provided for the sender")
+
```
Comment (liamlynch-data) [inline custom-recipes/send-mails-from-contacts-dataset/recipe.py:140]: Explaining patch changes: added validation of sender being present as it was missing

File: custom-recipes/send-mails-from-contacts-dataset/recipe.py [hunk 9]
```diff
@@ -34,25 +34,33 @@ def to_real_channel_id(channel_id):
...
+        recipients_list = recipients.split(",")
 
-# Validates that columns (eg recipient, cc, bcc) are string datatype
-def confirm_string_column(column_name, schema):
+    filter_out_blanks = lambda item: bool(item) and not (isinstance(item, (str)) and item.isspace())
+    filtered_list = list(filter(filter_out_blanks, recipients_list))
+    return filtered_list
+
+# Validates column exists in schema
+def confirm_column(column_name, schema):
```
Comment (liamlynch-data) [inline custom-recipes/send-mails-from-contacts-dataset/recipe.py]: Explaining patch changes: changed confirm_string_column to just confirm_column as:
1. We actually accept array types too
2. Even some non-string typed columns were working previously if they had the right value in, and we do not want to break any existing uses of the plugin that work

File: custom-recipes/send-mails-from-contacts-dataset/recipe.py [hunk 10]
```diff
@@ -187,8 +207,8 @@ def confirm_string_column(column_name, schema):
                 email_body_text = build_email_message_text(use_body_value, body_template, attachments_templating_dict, contact_dict, body_column,
                                                          use_html_body_value)
                 recipients = parse_recipients(recipients_string)
-                cc_recipients = parse_recipients(cc_string) if cc_string else [""]
-                bcc_recipients = parse_recipients(bcc_string) if bcc_string else [""]  
+                cc_recipients = parse_recipients(cc_string)
+                bcc_recipients = parse_recipients(bcc_string)
```
Comment (liamlynch-data) [inline custom-recipes/send-mails-from-contacts-dataset/recipe.py]: Explaining patch changes: defaulting to `[""]` breaks channel-based email clients. Now that we filter blank/empty values and also handle `None` in `parce_recipients()`, to give an empty array `[]` if necessary, we don't need the else part here, and I have tested that the empty array is handled well downstream

File: custom-recipes/send-mails-from-contacts-dataset/recipe.py [hunk 11]
```diff
@@ -200,7 +220,7 @@ def confirm_string_column(column_name, schema):
                 if writer:
                     writer.write_row_dict(contact_dict)
             except Exception as e:
-                logging.exception("Send failed")
+                logging.exception("Send failed: %s", str(e))
```
Comment (liamlynch-data) [inline custom-recipes/send-mails-from-contacts-dataset/recipe.py:223]: Explaining patch changes: may as well log the error as well as return it, makes debuging easier

File: python-lib/dku_email_client.py [hunk 12]
```diff
@@ -86,6 +86,9 @@ def send_email(self, sender, recipients, cc_recipients, bcc_recipients, email_su
         files = [(a.file_name, a.data, f"{a.mime_type}/{a.mime_subtype}") for a in attachment_files]
         sender_to_use = None if self.channel.sender else sender
 
+        if not recipients:
+            raise Exception("No recipients provided to send to")
+
```
Comment (liamlynch-data) [inline python-lib/dku_email_client.py:91]: Explaining patch changes: we get a much improved error message if we throw here - the error messages had become a bit cryptic due to the other changes

File: python-lib/dku_email_client.py [hunk 13]
```diff
@@ -169,10 +172,16 @@ def send_single_email(self, sender, recipients, cc_recipients, bcc_recipients, e
...
+        try:
+            self.smtp.sendmail(sender, all_recipients, msg.as_string())
+        finally:
+            # Explicitly reset the session or "Error: nested MAIL command" error is possible
+            self.smtp.rset()
 
     def send_email(self, sender, recipients, cc_recipients, bcc_recipients, email_subject, email_body, attachment_files):
         attachment_mimes = self.attachments_to_mime(attachment_files)
+        if not recipients:
+            raise Exception("No recipients provided to send to")
```
Comment (liamlynch-data) [inline python-lib/dku_email_client.py:184]: Explaining patch changes: again we get a much improved error message if we throw here - the error messages had become a bit cryptic due to the other changes

File: python-lib/dku_email_client.py [hunk 14]
```diff
@@ -169,10 +172,16 @@ def send_single_email(self, sender, recipients, cc_recipients, bcc_recipients, e
         for mime_app in attachment_mimes:
             msg.attach(mime_app)
 
-        self.smtp.sendmail(sender, all_recipients, msg.as_string())
+        try:
+            self.smtp.sendmail(sender, all_recipients, msg.as_string())
+        finally:
+            # Explicitly reset the session or "Error: nested MAIL command" error is possible
+            self.smtp.rset()
```
Comment (liamlynch-data) [inline python-lib/dku_email_client.py:179]: Explaining patch changes: with multiple recipients on the STMP client, with one failure the client would end up in a bad state and everything would fail. It seems likely this problem came in when the multiple recipients feature was added (i.e. it was not caused by the changes in this PR AFAIK)

**Comment by liamlynch-data** [comment]
LGTM and tested. The following tests were run with the patch code applied:

**Basic tests**

Default: Normal email, no CC/BCC, Email body in table (text), no attachments ✅ 

Add CC/BCC fields ✅ 
With Email arrays ✅ 
Subject in column (e.g. status) ✅ 
Sender in column / value ✅ 

**Error validation** 
Non-existent column for all the below:  ✅ 
Recipient Email
Sender
CC
BCC
Email body
Subject

Values:
No body - ✅ 
No subject - ✅ 
No sender - ✅ 

**Email body**
(Email body in table) 
Email body plain text - with a variable ✅ 
Email body html - with a variable ✅ 
Conditional formatting - SKIP - not relevant to change

**Attachments**
Excel attachment ✅ 
Tab delimited ✅ 
Comma delimited ✅ 

Email body plain text - with attach variable -  not relevant, SKIP
Email body html - with attach variable - not relevant, SKIP

**Regression** 
Recipe in project exported from an old version still works - with CSV tab delimited attachments ✅ 

**Channels** ✅ 
Repeated above via direct SMTP settings and a channel

**CC/BCC** ✅ 
Used an AWS SES mail channel and the dummy recipient address indicated in the messaging MTP AWS section as the sender to test CC and BCC emails (routed to addressed to me, an approved recipient) would get routed properly in the real world

**Review by liamlynch-data — APPROVED** [review]
Approving - see previous comment for details

**Comment by liamlynch-data** [comment]
@alexbourret @MayeulRousselet - as discussed it would be great if one of you could take a quick look, given the validation changes and the issues with multiple recipients we fixed

File: plugin.json [hunk 15]
```diff
@@ -1,6 +1,6 @@
 {
 	"id": "sendmail",
-	"version": "1.0.3",
+	"version": "1.0.4",
 	"meta": {
 		"label": "Send emails",
```
Comment (alexbourret) [inline plugin.json:5]: the tags should now be ["Productivity","Export"] and a `category` should be added in `meta`: `"category": "Digital Marketing",`

File: CHANGELOG.md [hunk 16]
```diff
@@ -1,5 +1,9 @@
 # Changelog
 
+## [Version 1.0.4](https://github.com/dataiku/dss-plugin-sendmail/releases/tag/v1.0.4) - Feature release - 2025-12
+
+- Add comma-delimited CSV option
+- Add cc and bcc fields
```
Comment (MayeulRousselet) [inline CHANGELOG.md]: Bump to a minor

File: custom-recipes/send-mails-from-contacts-dataset/recipe.json [hunk 17]
```diff
@@ -115,6 +106,28 @@
...
+            "columnRole" : "contacts",
+            "description" : "Recipient of the email (from a column)",
+            "mandatory": true
+        },
+        {
+            "name": "cc_column",
+            "label" : "cc (column)",
+            "type": "COLUMN",
+            "columnRole" : "contacts",
+            "description" : "Courtesy copy (from a column)"
```
Comment (alexbourret) [inline custom-recipes/send-mails-from-contacts-dataset/recipe.json]: Should it say `comma separated or list` or are we sure all cases are covered ?

File: custom-recipes/send-mails-from-contacts-dataset/recipe.json [hunk 18]
```diff
@@ -115,6 +106,28 @@
...
+        },
+        {
+            "name": "bcc_column",
+            "label" : "bcc (column)",
+            "type": "COLUMN",
+            "columnRole" : "contacts",
+            "description" : "Blind courtesy copy (from a column)"
```
Comment (alexbourret) [inline custom-recipes/send-mails-from-contacts-dataset/recipe.json]: Same as above

**Review by alexbourret — COMMENTED** [review]
LGTM, just update the plugin.json with the new tags and category, and maybe clarify the expected format in the cc and bcc columns

File: CHANGELOG.md (hunk 16 above)
Comment (liamlynch-data) [reply CHANGELOG.md]: You mean we should go to Version 1.1.0?

Comment (holstbone) [reply CHANGELOG.md]: Revised to v1.1.0

File: plugin.json (hunk 15 above)
Comment (holstbone) [reply plugin.json:5]: Revised to v1.1.0. Also adjusted tags, and added "category"

File: custom-recipes/send-mails-from-contacts-dataset/recipe.json (hunk 17 above)
Comment (holstbone) [reply custom-recipes/send-mails-from-contacts-dataset/recipe.json]: Revised "description" to read:  "Courtesy copy (comma-separated or array, from a column)"

File: custom-recipes/send-mails-from-contacts-dataset/recipe.json (hunk 18 above)
Comment (holstbone) [reply custom-recipes/send-mails-from-contacts-dataset/recipe.json]: Revised "description" to read: "Blind courtesy copy (comma-separated or array, from a column)"

**Review by RoyTeddy — CHANGES_REQUESTED** [review]
Regression found:
if Subject (column) or Body (column) is a datetime column we have this error:
`[18:05:25] [INFO] [dku.utils]  - TypeError: unsupported operand type(s) for +: 'datetime.datetime' and 'str'`
Works with previous version 1.0.3 in release Version: 14.3.2 here https://released-design.qa.managedinstances.dkucloud-dev.com/projects/TEDDY_COPY/flow/

error caused by the input dataset which contains an empty row so no bug here 😅

**Comment by RoyTeddy** [comment]
Found an issue but already there before: 
in the Body with custom value, if we specify a dataset name with numbers like {{ attachments.16rows.html_table }}
-> beautiful error
```
[2026/01/20-18:25:13.499] [null-err-72] [INFO] [dku.utils]  - *************** Recipe code failed **************
[2026/01/20-18:25:13.500] [null-err-72] [INFO] [dku.utils]  - Begin Python stack
[2026/01/20-18:25:13.500] [null-err-72] [INFO] [dku.utils]  - Traceback (most recent call last):
[2026/01/20-18:25:13.501] [null-err-72] [INFO] [dku.utils]  -   File "/Users/teddy.roy@dataiku.com/devenv/DSS14/dss-home/jobs/LEGACYCFFOREXPORT/Build_send_review__NP__2026-01-20T17-25-12.075/compute_send_review_NP/custom-python-recipe/pyout4nbErQ4gweO6/python-exec-wrapper.py", line 204, in <module>
[2026/01/20-18:25:13.501] [null-err-72] [INFO] [dku.utils]  -     exec(f.read())
[2026/01/20-18:25:13.501] [null-err-72] [INFO] [dku.utils]  -   File "<string>", line 166, in <module>
[2026/01/20-18:25:13.502] [null-err-72] [INFO] [dku.utils]  -   File "/Users/teddy.roy@dataiku.com/devenv/DSS14/dss-home/code-envs/python/plugin_sendmail_managed/lib/python3.9/site-packages/jinja2/environment.py", line 1092, in from_string
[2026/01/20-18:25:13.502] [null-err-72] [INFO] [dku.utils]  -     return cls.from_code(self, self.compile(source), gs, None)
[2026/01/20-18:25:13.503] [null-err-72] [INFO] [dku.utils]  -   File "/Users/teddy.roy@dataiku.com/devenv/DSS14/dss-home/code-envs/python/plugin_sendmail_managed/lib/python3.9/site-packages/jinja2/environment.py", line 757, in compile
[2026/01/20-18:25:13.503] [null-err-72] [INFO] [dku.utils]  -     self.handle_exception(source=source_hint)
[2026/01/20-18:25:13.503] [null-err-72] [INFO] [dku.utils]  -   File "/Users/teddy.roy@dataiku.com/devenv/DSS14/dss-home/code-envs/python/plugin_sendmail_managed/lib/python3.9/site-packages/jinja2/environment.py", line 925, in handle_exception
[2026/01/20-18:25:13.504] [null-err-72] [INFO] [dku.utils]  -     raise rewrite_traceback_stack(source=source)
[2026/01/20-18:25:13.504] [null-err-72] [INFO] [dku.utils]  -   File "<unknown>", line 4, in template
[2026/01/20-18:25:13.504] [null-err-72] [INFO] [dku.utils]  - jinja2.exceptions.TemplateSyntaxError: expected token 'end of print statement', got 'rows'
```

it is a Python limitation. 
Attributes that start with a number, such as "123property" have to be expressed as {{customer['123property']}}

so works with {{ attachments['16rows'].html_table }}
not a bug 😅

**Review by RoyTeddy — APPROVED** [review]
LGTM and tested 👍
Conditional formatting works:
- in attachements ✅
- in body ✅

**Comment by liamlynch-data** [comment]
> Found an issue but already there before: in the Body with custom value, if we specify a dataset name with numbers like {{ attachments.16rows.html_table }} -> beautiful error
> 
> ```
...
> ```
> 
> it is a Python limitation. Attributes that start with a number, such as "123property" have to be expressed as {{customer['123property']}}
> 
> so works with {{ attachments['16rows'].html_table }} not a bug 😅

This one seems pre-existing, nothing to do with the change, and kind of as designed (or a limitation we can't do much about) - so we can totally ignore this can't we @RoyTeddy  ?
(Thanks for the thorough testing though!)

**Comment by RoyTeddy** [comment]
> > Found an issue but already there before: in the Body with custom value, if we specify a dataset name with numbers like {{ attachments.16rows.html_table }} -> beautiful error
> 
> ...
> 
> > ```
> > 
> > it is a Python limitation. Attributes that start with a number, such as "123property" have to be expressed as {{customer['123property']}}
> > 
> > so works with {{ attachments['16rows'].html_table }} not a bug 😅
> > ```
> 
> This one seems pre-existing, nothing to do with the change, and kind of as designed (or a limitation we can't do much about) - so we can totally ignore this can't we @RoyTeddy ? (Thanks for the thorough testing though!)

@liamlynch-data yes we can ignore it 😉

**Comment by liamlynch-data** [comment]
> Regression found: if Subject (column) or Body (column) is a datetime column we have this error: `[18:05:25] [INFO] [dku.utils] - TypeError: unsupported operand type(s) for +: 'datetime.datetime' and 'str'` Works with previous version 1.0.3 in release Version: 14.3.2 here https://released-design.qa.managedinstances.dkucloud-dev.com/projects/TEDDY_COPY/flow/
> 
> error caused by the input dataset which contains an empty row so no bug here 😅

Yes this is not a regression, v 1.0.3 of the plugin does this, e.g.:
https://staging-design.qa.managedinstances.dkucloud-dev.com/projects/SENDMAIL1/datasets/ergergerg/explore/

image

**Review by MayeulRousselet — CHANGES_REQUESTED** [review]
Arriving late to the party, but I have a fundamental issue: If there are multiple recipient in the same row, and 1 CC/BCC, the CC/BCC will get spammed as he receive one email per recipient.

2 solutions:
- Cheapest: having a warning next to `Recipient` and `CC/BCC`
- Hardest, but better: havnig the possibility to send the same email to multiple recipient (need a DSS change)

**Comment by liamlynch-data** [comment]
> Arriving late to the party, but I have a fundamental issue: If there are multiple recipient in the same row, and 1 CC/BCC, the CC/BCC will get spammed as he receive one email per recipient.
> 
> 2 solutions:
> 
> * Cheapest: having a warning next to `Recipient` and `CC/BCC`
> * Hardest, but better: havnig the possibility to send the same email to multiple recipient (need a DSS change)

I think we need to do the cheap option for now - @holstbone would you be able to add the warning? If you're not sure what needs doing I could send you a patch, but it should be simple 

Re " the possibility to send the same email to multiple recipient" - we have a card already for this, but it's not scheduled:
https://app.shortcut.com/dataiku/story/233246/sendmail-add-an-option-so-if-multiple-senders-they-all-get-the-same-email

I'd also like to reiterate that we should really fix the following issue before we release with this (not necessarily in the same PR but before we release a version of the plugin): 
https://app.shortcut.com/dataiku/story/298155/sendmail-plugin-wrong-column-type-for-email-fields-could-be-handled-better

**Comment by MayeulRousselet** [comment]
> > Arriving late to the party, but I have a fundamental issue: If there are multiple recipient in the same row, and 1 CC/BCC, the CC/BCC will get spammed as he receive one email per recipient.
> > 2 solutions:
> > 
> > * Cheapest: having a warning next to `Recipient` and `CC/BCC`
> > * Hardest, but better: havnig the possibility to send the same email to multiple recipient (need a DSS change)
> 
> I think we need to do the cheap option for now - @holstbone would you be able to add the warning? If you're not sure what needs doing I could send you a patch, but it should be simple
> 
> Re " the possibility to send the same email to multiple recipient" - we have a card already for this, but it's not scheduled: https://app.shortcut.com/dataiku/story/233246/sendmail-add-an-option-so-if-multiple-senders-they-all-get-the-same-email
> 
> I'd also like to reiterate that we should really fix the following issue before we release with this (not necessarily in the same PR but before we release a version of the plugin): https://app.shortcut.com/dataiku/story/298155/sendmail-plugin-wrong-column-type-for-email-fields-could-be-handled-better

Logged as a blocker in SH @liamlynch-data

**Comment by holstbone** [comment]
Hi @MayeulRousselet and @liamlynch-data  I have updated the descriptions for cc_column and bcc_column to include a note about duplicate data causing multiple emails. 

Added text:  "**Note: Duplicate data will cause multiple emails.**"

Hopefully this addresses your concern!

 It looks like this now:
Screenshot 2026-04-15 at 7 21 47 PM

**Comment by holstbone** [comment]
Hi @MayeulRousselet - I wanted to follow up on two items:

1.) Does my adjustment made on April 15th address your concern? See: https://github.com/dataiku/dss-plugin-sendmail/pull/21#issuecomment-4256739374

2.) Regarding Shortcut #298155, did you actually test on this Pull Request (with the new code)? We did some better handling of types, but not certain if it specifically addresses the issue in the Shortcut. See: https://app.shortcut.com/dataiku/story/298155/sendmail-plugin-wrong-column-type-for-email-fields-could-be-handled-better

**Comment by liamlynch-data** [comment]
> Hi @MayeulRousselet - I wanted to follow up on two items:
> 
> 1.) Does my adjustment made on April 15th address your concern? See: [#21 (comment)](https://github.com/dataiku/dss-plugin-sendmail/pull/21#issuecomment-4256739374)
> 
> 2.) Regarding Shortcut #298155, did you actually test on this Pull Request (with the new code)? We did some better handling of types, but not certain if it specifically addresses the issue in the Shortcut. See: https://app.shortcut.com/dataiku/story/298155/sendmail-plugin-wrong-column-type-for-email-fields-could-be-handled-better

Re 2) - I raised this after testing this PR, so this is something we still need to fix. But I can do it after this one is merged, no worries.
